show the right image name in job log and docker ps

liusdu opened the pull request:
### introduction

It is a new version of PR #10479 . 

> When we tag an Image with several names and we run one of them,
> The "create" job will log this event with
> +job log(create, containerID, Imagename).
> And the "Imagename" is always the first one (sorted). It is the
> same to "start/stop/rm" jobs. So use the correct name instand.

 I have discuss this with @LK4D4. We all think we should show just as what we type. He afraid that this may introduce break change. @tiborvass thinks we can show ubuntu:latest in the first case and 8eaa4ff0 in the second case. 

But  because when we start a container with Image ID. We cannot record a related image:tag in log.
So this will introduce a break change(not image:tag format).  So my patched does following.
### Image List

---

| Image tag | ImageID |
| --- | --- |
| a-ubuntu:latest | 8eaa4ff0 |
| ubuntu:latest | 8eaa4ff0 |
| z-ubunutu:latest | 8eaa4ff0 |
### behavior before  this patch series

---

| No | cmd | behaviour |
| --- | --- | --- |
| 1 | docker run -ti ubuntu bash | +jog(create, containerid, a-ubuntu:latest) |
| 3 | docker run -ti ubuntu:latest bash | +jog(create, containerid, a-ubuntu:latest) |
| 2 | docker run -ti 8eaa4ff0 bash | +jog(create, containerid, a-ubuntu:latest) |

---
### behavior  after this patch series

---

| No | cmd | behaviour |
| --- | --- | --- |
| 1 | docker run -ti ubuntu bash | +jog(create, containerid, ubuntu) |
| 3 | docker run -ti ubuntu:latest bash | +jog(create, containerid, ubuntu:latest) |
| 2 | docker run -ti 8eaa4ff0 bash | +jog(create, containerid, 8eaa4ff0) |

---
### extra changes
- Because the daemon log behavior has changed. It affects "docker events", which was used several times in ci. So I also fix these issues.
- docker ps also has similar issues. Fix them

aarch64 committed:
Show the right image name/ID in job log

When we tag an Image with several names and we run one of them,
The "create" job will log this event with
 +job log(create, containerID, Imagename).

And the "Imagename" is always the first one (sorted). It is the
same to "start/stop/rm" jobs. So use the correct name instand.
This PR refer to #10479

Signed-off-by: Liu Hua <sdu.liu@huawei.com>

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "m" git@github.com:liusdu/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~3
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

LK4D4 commented:
I'm okay with this.
Could you pls squash fixup commits?

liusdu commented:
hi @LK4D4 done， Please help me review

LK4D4 commented on daemon/list.go:
```diff
@@ -130,10 +127,6 @@ func (daemon *Daemon) Containers(job *engine.Job) error {
-		if tag == "" {
-			img = utils.ImageReference(img, graph.DEFAULTTAG)
-		}
 		out.SetJson("Image", img)
```
can be just `container.Config.Image` now.

LK4D4 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -292,6 +293,60 @@ func TestDaemonLoggingLevel(t *testing.T) {
 }
 
+func TestDaemonLoggingContent(t *testing.T) {
+
```
Would you mind to remove empty lines after `{` and before `}`?

liusdu replied:
Thanks ! @LK4D4

tiborvass commented:
@liusdu so you're basically agreeing with my comment https://github.com/docker/docker/pull/10479#issuecomment-81958550 ? Just wanna make sure :)

liusdu commented:
@tiborvass Thanks to you reply. I am afraid of this scenario: one starts two containers with busybox and busybox:latest.  With this patch docker events can identify them using flag "--filter".  But if we add default tag for busybox, we can not do it.

LK4D4 commented:
LGTM

tiborvass commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
 }
 
+func TestDaemonLoggingContent(t *testing.T) {
+	CheckImageNameInLog := func(d *Daemon, ImageName string) {
```
should start with small letters: `checkImageNameInLog` and `imageName`

tiborvass commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
+		}
+		content, _ := ioutil.ReadFile(d.logFile.Name())
+
+		reg := regexp.MustCompile(`\+job log\(create, (?P<ID>[0-9|a-z]*), (?P<ImageName>.*)\)`)
```
you should be able to declare this outside the function

tiborvass commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -292,6 +293,58 @@ func TestDaemonLoggingLevel(t *testing.T) {
+	}
+
+	// check Image name in job log
+	ImageNames := []string{"a-busybox", "b-busybox", "c-busybox"}
```
`imageNames`

tiborvass commented:
@liusdu couple of style changes, and it's good otherwise :)

liusdu commented:
Hi,  @tiborvass all done.

aarch64 committed:
fix up Image-name related issues in docker ps and CI

This patch include the following fixs:
 - fix image name error when docker ps
 - fix docker events test failure: use the exact image name for filter
 - fix docker build CI test failure due to "docker events" change

Because of change of daemon log behavior. Now we record
the exact Image name as you typed. So docker run -d busybux sh
and docker run -d busybox:latest are not the same in the log.
So it will affect the docker events. So change the related CI

Signed-off-by: Liu Hua <sdu.liu@huawei.com>

liusdu commented:
I add a bad patch to let event test pass .
in commit 6b5e2088 I sleep one second to let container stops. It is very bad.

But now I have a better idea. I can directly stop containers,not just wait. So related tests should be changed acorrdingly.

liusdu commented:
Pr #11981 also encounters this problem.

LK4D4 commented:
@liusdu It is fixed in #11949 (I hope at least).

tiborvass commented:
@LK4D4 I can confirm it was not fixed by #11949. I still see `busybox:latest` in all cases.

tiborvass commented:
LGTM

LK4D4 commented:
@tiborvass I meant event tests fail.

LK4D4 commented:
@liusdu Could you remove your patch to events test and rebase against master?

liusdu commented:
@LK4D4  @tiborvass  done

tiborvass commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -291,6 +292,57 @@ func TestDaemonLoggingLevel(t *testing.T) {
 
 	logDone("daemon - Logging Level")
 }
+func checkImageNameInLog(d *Daemon, imageName string) (error, string) {
```
@liusdu sorry you should keep this inside the test function like so:

```Go
var checkImageNameInLog = func(...) (...) {
    // ...
}
```

LK4D4 replied:
or even:
```
checkImageNameInLog := func(...) (...) {
    // ...
}
```

liusdu commented:
@LK4D4  @tiborvass  Good advice, Thank you! done

jfrazelle commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -292,6 +293,57 @@ func TestDaemonLoggingLevel(t *testing.T) {
+			t.Fatal(err, out)
+		}
+	}
+	d.Stop()
```
can we defer d.Stop() on line 324

liusdu replied:
Sorry for my bad code style!

LK4D4 commented:
LGTM

jfrazelle commented:
ah man needs a rebase so sorry

liusdu commented:
@jfrazelle  all fixed!

jfrazelle commented:
I dont know if that worked, it's still saying unable to merge due to conflicts? was i too slow :(

liusdu commented:
@jfrazelle  Yesterday I fixed conflicts with two commits. Today one more~~.  Codes related with this patch is in heave developing.  And after commit c9eb37f9., we do not the test for daemon log.

jfrazelle commented:
LGTM so sorry!

LK4D4 commented:
Ah, nice!

relistan commented:
This change unfortunately broke `docker ps` output as shown in #12595. The breaking change is here: https://github.com/liusdu/docker/commit/645c020f5ab7119cd06d7c6a790e9d99fe1cd309#diff-9320049615863c9b6ab00c749bbb2f09L139
